pwndbg/gdblib/kernel/kallsyms.py

```diff
@@ -387,15 +398,143 @@ def parse_symbol_table(self):
+        Original Source: https://github.com/marin-m/vmlinux-to-elf/blob/master/vmlinux_to_elf/kallsyms_finder.py
+        """
+        position = self.end_of_kallsyms_names_uncompressed
+        position += -position % 4
+
+        max_number_of_space_between_two_nulls = 0
+
+        # Go just after the first chunk of non-null bytes
 
-        for num_token in range(256):
-            token = ""
-            while self.kernel_ro_mem[position]:
-                token += chr(self.kernel_ro_mem[position])
+        # while position + 1 < len(self.kernel_img) and self.kernel_img[position + 1] == 0:
+        #     position += 1
+
+        while position + 1 < len(self.kernel_ro_mem) and self.kernel_ro_mem[position + 1] == 0:
+            position += 1
+
+        for null_separated_bytes_chunks in range(20):
+            num_non_null_bytes = 1  # we always start at a non-null byte in this loop
+            num_null_bytes = (
+                1  # we will at least encounter one null byte before the end of this loop
+            )
+
+            while True:
                 position += 1
+                assert position >= 0
+
+                if self.kernel_ro_mem[position] == 0:
+                    break
+                num_non_null_bytes += 1
+
+            while True:
+                position += 1
+                assert position >= 0
+
+                if self.kernel_ro_mem[position] != 0:
+                    break
+                num_null_bytes += 1
+
+            max_number_of_space_between_two_nulls = max(
+                max_number_of_space_between_two_nulls, num_non_null_bytes + num_null_bytes
+            )
+
+        if (
+            max_number_of_space_between_two_nulls % 2 == 1
+        ):  # There may be a leap to a shorter offset in the latest processed entries
+            max_number_of_space_between_two_nulls -= 1
+
+        if max_number_of_space_between_two_nulls not in (2, 4, 8):
+            print(M.error("Could not guess the architecture register size for kernel"))
+            return None
+
+        self.offset_table_element_size = max_number_of_space_between_two_nulls
+
+        # Once the size of a long has been guessed, use it to find
+        # the first offset (0)
+
+        position = self.end_of_kallsyms_names_uncompressed
+        position += -position % 4
+
+        # Go just at the first non-null byte
+        while position < len(self.kernel_ro_mem) and self.kernel_ro_mem[position] == 0:
             position += 1
-            tokens.append(token)
 
-        return tokens
+        likely_is_big_endian = position % self.offset_table_element_size > 1
+        if self.is_big_endian is None:  # Manual architecture specification
+            self.is_big_endian = likely_is_big_endian
+
+        if position % self.offset_table_element_size == 0:
+            position += self.offset_table_element_size
+        else:
+            position += -position + self.offset_table_element_size
```

Comment:
Is this correct? This is literally doing... `position = self.offset_table_element_size` iiuc?